plugins/source/scheduler_round_robin.go

```diff
@@ -0,0 +1,99 @@
+package source
+
+import (
+	"context"
+	"sync"
+
+	"github.com/cloudquery/plugin-sdk/schema"
+	"github.com/cloudquery/plugin-sdk/specs"
+	"golang.org/x/sync/semaphore"
+)
+
+func (p *Plugin) syncRoundRobin(ctx context.Context, spec specs.Source, client schema.ClientMeta, tables schema.Tables, resolvedResources chan<- *schema.Resource) {
+	tableConcurrency := max(spec.Concurrency/minResourceConcurrency, minTableConcurrency)
+	resourceConcurrency := tableConcurrency * minResourceConcurrency
+
+	p.tableSems = make([]*semaphore.Weighted, p.maxDepth)
+	for i := uint64(0); i < p.maxDepth; i++ {
+		p.tableSems[i] = semaphore.NewWeighted(int64(tableConcurrency))
+		// reduce table concurrency logarithmically for every depth level
+		tableConcurrency = max(tableConcurrency/2, minTableConcurrency)
+	}
+	p.resourceSem = semaphore.NewWeighted(int64(resourceConcurrency))
+
+	// we have this because plugins can return sometimes clients in a random way which will cause
+	// differences between this run and the next one.
+	preInitialisedClients := make([][]schema.ClientMeta, len(tables))
+	for i, table := range tables {
+		clients := []schema.ClientMeta{client}
+		if table.Multiplex != nil {
+			clients = table.Multiplex(client)
+		}
+		preInitialisedClients[i] = clients
+		// we do this here to avoid locks so we initial the metrics structure once in the main goroutines
+		// and then we can just read from it in the other goroutines concurrently given we are not writing to it.
+		p.metrics.initWithClients(table, clients)
+	}
+
+	// We start a goroutine that logs the metrics periodically.
+	// It needs its own waitgroup
+	var logWg sync.WaitGroup
+	logWg.Add(1)
+
+	logCtx, logCancel := context.WithCancel(ctx)
+	go p.periodicMetricLogger(logCtx, &logWg)
+
+	var wg sync.WaitGroup
+	type tableClient struct {
+		table  *schema.Table
+		client schema.ClientMeta
+	}
+
+	// interleave table-clients so that we get:
+	// table1-client1, table2-client1, table3-client1, table1-client2, table2-client2, table3-client2, ...
+	tableClients := make([]tableClient, 0)
+	c := 0
+	for {
+		addedNew := false
+		for i, table := range tables {
+			if c < len(preInitialisedClients[i]) {
+				tableClients = append(tableClients, tableClient{table: table, client: preInitialisedClients[i][c]})
+				addedNew = true
+			}
+		}
+		c++
+		if !addedNew {
+			break
+		}
+	}
+
+	for _, tc := range tableClients {
+		table := tc.table
+		cl := tc.client
+		if err := p.tableSems[0].Acquire(ctx, 1); err != nil {
+			// This means context was cancelled
+			wg.Wait()
+			// gracefully shut down the logger goroutine
+			logCancel()
+			logWg.Wait()
+			return
+		}
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			defer p.tableSems[0].Release(1)
```

Comment:
[nit]: Is lookup for `p.tableSems[0]` necessary, especially, in a loop? Shouldn't it be extracted?

Reply:
Not sure, but it seems like a micro-optimization and not relevant to this PR, as this is not unique to the round robin scheduler. Feel free to try it and see if it makes a difference in the benchmarks.